File: codility/CountSemiprimes.js
function solution(N, P, Q) {
    // Implement your solution here
    const primes = getPrime(N)
    const semiArr = [];
    for (let i = 0; i < primes.length; i++) {
        const p1 = primes[i]
        for (let j = 0; j < primes.length; j++) {
            const p2 = primes[j]
            if (p1 * p2 > N) {
                break;
            }
            semiArr[p1 * p2] = true;
        }
    }
    // console.log(semiArr)
    const result = [];

    const count = new Array(N + 1);
    for (let i = 0; i < N + 1; i++) {
        if (i === 0) {
            count[i] = 0
        } else {
            count[i] = count[i - 1]
        }
        if (semiArr[i]) {
            count[i]++;
        }

    }
    // console.log(count)
    for (let i = 0; i < P.length; i++) {
        const p = P[i], q = Q[i];
        let gap = count[q] - count[p];
        if(semiArr[p]){gap++;}
        result.push(gap)
    }

    return result;

    function getPrime(n) {
        const arr = new Array(n + 1).fill(1);
        arr[0] = 0;
        arr[1] = 0;
        const max = Math.floor(Math.sqrt(n));
        for (let i = 2; i <= max; i++) {
            if (arr[i] === 0) { continue; }
            for (let j = i + 1; j <= n; j++) {
                if (j % i === 0) {
                    arr[j] = 0;
                }
            }
        }
        const result = [];
        for (let i = 0; i < arr.length; i++) {
            if (arr[i] === 1) {
                result.push(i)
            }
        }
        return result;
    }
}

File: codility/python/CountSemiprimes.py
def solution(N, P, Q):
    # Implement your solution here
    primes = get_prime(int(N/2))
    # print(primes)
    semiprimes = []
    for i in range(len(primes)):
        prime = primes[i]
        j = i
        for j in range(i, len(primes)):
            if prime * primes[j] > N:
                break
            semiprimes.append(prime * primes[j])
    semiprimes.sort()
    # print(semiprimes)
    counter = [0] * (N+1)
    index = 0
    count = 0
    for i in range(1, N+1):
        if index < len(semiprimes) and i == semiprimes[index]:
            index += 1
            count += 1
        counter[i] = count
    # print(counter)
    result = []
    for i in range(len(P)):
        p = P[i]
        q = Q[i]
        result.append(counter[q] - counter[p-1])
    return result
    
def get_prime(n):
    primes=[1] * (n+1)
    # max_num = int(n ** 0.5) + 1
    

    for i in range(2, n+1):
        d = 2
        while d * d <= i:
            if i % d == 0:
                primes[i] = 0
                break
            d += 1
    result = []
    for i in range(2, len(primes)):
        if primes[i] == 1:
            result.append(i)
    return result